File: backend/src/services/GameEngine.ts
export enum CardColor {
  Red = 'red',
  Blue = 'blue',
  Green = 'green',
  Yellow = 'yellow',
  Wild = 'wild',
}

export enum CardType {
  Number = 'number',
  Skip = 'skip',
  Reverse = 'reverse',
  DrawTwo = 'drawTwo',
  Wild = 'wild',
  WildDrawFour = 'wildDrawFour',
  CustomBlank = 'customBlank',
}

export interface Card {
  id: string;
  color: CardColor;
  type: CardType;
  number?: number;
  customRule?: string;
}

export interface GamePlayer {
  id: string;
  name: string;
  socketId: string;
  hand: Card[];
  score: number;
  hasCalledUno: boolean;
  isConnected: boolean;
}

export interface GameState {
  id: string;
  roomId: string;
  status: 'waiting' | 'inProgress' | 'finished';
  players: GamePlayer[];
  currentPlayerIndex: number;
  topCard: Card | null;
  direction: 'clockwise' | 'counterClockwise';
  canChallenge: boolean;
  createdAt: Date;
  finishedAt?: Date;
}

export class GameEngine {
  private static readonly STARTING_HAND = 7;

  /**
   * Create a standard UNO deck
   */
  static createDeck(): Card[] {
    const cards: Card[] = [];
    let id = 0;

    const colors = [CardColor.Red, CardColor.Blue, CardColor.Green, CardColor.Yellow];

    // Number cards
    for (const color of colors) {
      // 0 appears once
      cards.push({
        id: `card_${id++}`,
        color,
        type: CardType.Number,
        number: 0,
      });

      // 1-9 appear twice
      for (let number = 1; number <= 9; number++) {
        for (let i = 0; i < 2; i++) {
          cards.push({
            id: `card_${id++}`,
            color,
            type: CardType.Number,
            number,
          });
        }
      }

      // Action cards (2 each)
      for (let i = 0; i < 2; i++) {
        cards.push({
          id: `card_${id++}`,
          color,
          type: CardType.Skip,
        });
        cards.push({
          id: `card_${id++}`,
          color,
          type: CardType.Reverse,
        });
        cards.push({
          id: `card_${id++}`,
          color,
          type: CardType.DrawTwo,
        });
      }
    }

    // Wild cards (4 of each)
    for (let i = 0; i < 4; i++) {
      cards.push({
        id: `card_${id++}`,
        color: CardColor.Wild,
        type: CardType.Wild,
      });
      cards.push({
        id: `card_${id++}`,
        color: CardColor.Wild,
        type: CardType.WildDrawFour,
      });
    }

    // Custom blank cards (4)
    for (let i = 0; i < 4; i++) {
      cards.push({
        id: `card_${id++}`,
        color: CardColor.Wild,
        type: CardType.CustomBlank,
      });
    }

    return cards;
  }

  /**
   * Shuffle deck using Fisher-Yates algorithm
   */
  static shuffleDeck(cards: Card[]): Card[] {
    const deck = [...cards];
    for (let i = deck.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1));
      [deck[i], deck[j]] = [deck[j], deck[i]];
    }
    return deck;
  }

  /**
   * Validate if a card can be played
   */
  static isValidMove(
    card: Card,
    topCard: Card | null,
    activeColor: CardColor | null
  ): boolean {
    if (topCard === null) return true;

    // Wild cards can always be played
    if (card.color === CardColor.Wild) {
      return true;
    }

    // Match color with active color (from Wild card)
    if (activeColor && activeColor !== CardColor.Wild && card.color === activeColor) {
      return true;
    }

    // Match color
    if (card.color === topCard.color) {
      return true;
    }

    // Match type or number
    if (card.type === topCard.type) {
      return true;
    }

    if (
      card.type === CardType.Number &&
      topCard.type === CardType.Number &&
      card.number === topCard.number
    ) {
      return true;
    }

    return false;
  }

  /**
   * Get playable cards from hand
   */
  static getPlayableCards(
    hand: Card[],
    topCard: Card | null,
    activeColor: CardColor | null
  ): Card[] {
    return hand.filter((card) =>
      this.isValidMove(card, topCard, activeColor)
    );
  }

  /**
   * Validate Wild Draw Four legality
   */
  static isWildDrawFourLegal(
    playerHand: Card[],
    declaredColor: CardColor | null
  ): boolean {
    if (!declaredColor || declaredColor === CardColor.Wild) {
      return false;
    }

    // Check if player has any cards of the declared color (excluding wild)
    return !playerHand.some(
      (card) => card.color === declaredColor
    );
  }

  /**
   * Calculate score for cards in hand
   */
  static calculateHandScore(cards: Card[]): number {
    return cards.reduce((total, card) => {
      switch (card.type) {
        case CardType.Number:
          return total + (card.number || 0);
        case CardType.Skip:
        case CardType.Reverse:
        case CardType.DrawTwo:
          return total + 20;
        case CardType.Wild:
        case CardType.WildDrawFour:
        case CardType.CustomBlank:
          return total + 50;
        default:
          return total;
      }
    }, 0);
  }

  /**
   * Calculate round scores
   */
  static calculateRoundScores(
    playerHands: Map<string, Card[]>,
    winnerId: string
  ): Map<string, number> {
    const scores = new Map<string, number>();

    for (const [playerId, hand] of playerHands) {
      if (playerId === winnerId) {
        scores.set(playerId, 0);
      } else {
        scores.set(playerId, this.calculateHandScore(hand));
      }
    }

    return scores;
  }

  /**
   * Get next player index
   */
  static getNextPlayerIndex(
    currentIndex: number,
    totalPlayers: number,
    direction: 'clockwise' | 'counterClockwise',
    skipCount: number = 0
  ): number {
    if (totalPlayers <= 1) return 0;

    let nextIndex = currentIndex;
    const steps = skipCount + 1;

    if (direction === 'clockwise') {
      nextIndex = (currentIndex + steps) % totalPlayers;
    } else {
      nextIndex = (currentIndex - steps + totalPlayers * steps) % totalPlayers;
    }

    return nextIndex;
  }

  /**
   * Check if special rule applies for 1v1 (Reverse becomes Skip)
   */
  static isHeadsUpGame(playerCount: number): boolean {
    return playerCount === 2;
  }

  /**
   * Validate deck integrity
   */
  static validateDeck(cards: Card[]): boolean {
    // Check size
    if (cards.length < 108 || cards.length > 112) {
      return false;
    }

    // Check no duplicates
    const ids = new Set(cards.map((c) => c.id));
    return ids.size === cards.length;
  }

  /**
   * Deal initial hands
   */
  static dealInitialHands(
    deck: Card[],
    playerCount: number
  ): Map<string, Card[]> {
    const hands = new Map<string, Card[]>();
    let cardIndex = 0;

    for (let i = 0; i < playerCount; i++) {
      const hand: Card[] = [];
      for (let j = 0; j < this.STARTING_HAND; j++) {
        if (cardIndex < deck.length) {
          hand.push(deck[cardIndex++]);
        }
      }
      hands.set(`player_${i}`, hand);
    }

    return hands;
  }
}

File: flutter_app/lib/engine/game_engine.dart
import 'package:uno_game/models/card_model.dart';

class GameEngine {
  /// Check if a card can be played on the discard pile
  static bool isValidMove(Card card, Card? topCard, CardColor? activeColor) {
    if (topCard == null) return true;

    // Wild cards can always be played
    if (card.color == CardColor.wild) {
      return true;
    }

    // Check color match (for wild cards that chose a color)
    if (activeColor != null && activeColor != CardColor.wild) {
      if (card.color == activeColor) return true;
    }

    // Check color match (direct)
    if (card.color == topCard.color) return true;

    // Check number or type match
    if (card.type == topCard.type) return true;
    if (card.type == CardType.number && topCard.type == CardType.number) {
      return card.number == topCard.number;
    }

    return false;
  }

  /// Check if a Wild Draw Four is played illegally
  static bool isWildDrawFourChallengeable(
    List<Card> playerHand,
    CardColor? declaredColor,
  ) {
    // If no color was declared, can't challenge
    if (declaredColor == null || declaredColor == CardColor.wild) {
      return false;
    }

    // Check if player has any cards matching the declared color
    final hasCardOfColor = playerHand.any((card) {
      return card.color == declaredColor && card.color != CardColor.wild;
    });

    // If player has a card of the declared color, they played illegally
    return hasCardOfColor;
  }

  /// Get cards that can be played (client-side validation hint)
  static List<Card> getPlayableCards(
    List<Card> hand,
    Card? topCard,
    CardColor? activeColor,
  ) {
    if (topCard == null) return hand;
    return hand
        .where((card) => isValidMove(card, topCard, activeColor))
        .toList();
  }

  /// Calculate round score for all players
  static Map<String, int> calculateRoundScore(
    Map<String, List<Card>> playerHands,
    String winnerId,
  ) {
    final scores = <String, int>{};
    
    for (final entry in playerHands.entries) {
      if (entry.key == winnerId) {
        scores[entry.key] = 0;
      } else {
        int handValue = 0;
        for (final card in entry.value) {
          handValue += card.getPointValue();
        }
        scores[entry.key] = handValue;
      }
    }
    
    return scores;
  }

  /// Get the number of cards a player should draw
  static int getDrawCardsForAction(
    List<Card> lastPlayedCards,
    bool allowStacking,
  ) {
    if (!allowStacking) {
      // Just the card effect
      if (lastPlayedCards.isEmpty) return 0;
      return lastPlayedCards.last.type == CardType.drawTwo ? 2 : 4;
    }

    // Stacking enabled - count all draw cards
    int totalDraw = 0;
    for (final card in lastPlayedCards) {
      if (card.type == CardType.drawTwo) totalDraw += 2;
      if (card.type == CardType.wildDrawFour) totalDraw += 4;
    }
    return totalDraw;
  }

  /// Validate deck consistency
  static bool validateDeckIntegrity(List<Card> deck) {
    // Check total card count (should be 108-112)
    if (deck.length < 108 || deck.length > 112) return false;

    // Check no duplicates
    final ids = deck.map((c) => c.id).toSet();
    if (ids.length != deck.length) return false;

    return true;
  }

  /// Shuffle a deck using Fisher-Yates algorithm
  static List<Card> shuffleDeck(List<Card> deck) {
    final shuffled = List<Card>.from(deck);
    final random = DateTime.now().microsecond; // Poor man's random seed
    
    for (int i = shuffled.length - 1; i > 0; i--) {
      final j = (random * (i + 1)).toInt() % (i + 1);
      final temp = shuffled[i];
      shuffled[i] = shuffled[j];
      shuffled[j] = temp;
    }
    
    return shuffled;
  }

  /// Deal initial hands
  static Map<String, List<Card>> dealInitialHands(
    List<Card> deck,
    List<String> playerIds,
    int cardsPerPlayer,
  ) {
    final hands = <String, List<Card>>{};
    int cardIndex = 0;

    for (final playerId in playerIds) {
      final hand = <Card>[];
      for (int i = 0; i < cardsPerPlayer; i++) {
        if (cardIndex < deck.length) {
          hand.add(deck[cardIndex++]);
        }
      }
      hands[playerId] = hand;
    }

    return hands;
  }

  /// Check if a player should be forced to draw (no valid moves)
  static bool hasValidMove(
    List<Card> hand,
    Card topCard,
    CardColor? activeColor,
  ) {
    return getPlayableCards(hand, topCard, activeColor).isNotEmpty;
  }

  /// Get next player considering skips and reverses
  static int getNextPlayerIndex(
    int currentIndex,
    int totalPlayers,
    bool reverse,
    int skipCount,
  ) {
    if (totalPlayers <= 1) return 0;

    int nextIndex = currentIndex;
    for (int i = 0; i < skipCount + 1; i++) {
      if (reverse) {
        nextIndex = (nextIndex - 1 + totalPlayers) % totalPlayers;
      } else {
        nextIndex = (nextIndex + 1) % totalPlayers;
      }
    }
    return nextIndex;
  }
}

/// Event types for game progress
enum GameEventType {
  cardPlayed,
  cardDrawn,
  playerSkipped,
  directionChanged,
  unoCall,
  challengeIssued,
  challengeResolved,
  roundEnded,
  gameEnded,
}

class GameEvent {
  final GameEventType type;
  final String playerId;
  final Map<String, dynamic> data;
  final DateTime timestamp;

  GameEvent({
    required this.type,
    required this.playerId,
    this.data = const {},
    DateTime? timestamp,
  }) : timestamp = timestamp ?? DateTime.now();

  Map<String, dynamic> toJson() {
    return {
      'type': type.toString().split('.').last,
      'playerId': playerId,
      'data': data,
      'timestamp': timestamp.toIso8601String(),
    };
  }
}
